engine/src/test/java/org/camunda/bpm/engine/test/bpmn/tasklistener/TaskListenerEventLifecycleTest.java

```diff
@@ -0,0 +1,586 @@
+    Task task = taskService.createTaskQuery().singleResult();
+
+    // when
+    taskService.setAssignee(task.getId(), "gonzo");
+
+    // then
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    assertThat(orderedEvents.size()).isEqualTo(3);
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_UPDATE,
+                                              TaskListener.EVENTNAME_ASSIGNMENT);
+  }
+
+  @Test
+  public void shouldFireCompleteEventOnTaskCompletedInUpdateListener() {
+    // given
+    BpmnModelInstance model = createModelWithTaskEventsRecorderOnAssignedUserTask(TRACKED_EVENTS,
+                                                                                  null,
+                                                                                  TaskListener.EVENTNAME_UPDATE,
+                                                                                  CompletingTaskListener.class);
+    testRule.deploy(model);
+    runtimeService.startProcessInstanceByKey("process");
+    Task task = taskService.createTaskQuery().singleResult();
+
+    // when
+    taskService.setPriority(task.getId(), 3000);
+
+    // then
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    // assignment event should not be processed
+    assertThat(orderedEvents.size()).isEqualTo(3);
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_UPDATE,
+                                              TaskListener.EVENTNAME_COMPLETE);
+  }
+
+  @Test
+  public void shouldFireCompleteEventOnTaskCompletedInAssignmentListenerWhenTaskUpdated() {
+    // given
+    BpmnModelInstance model = createModelWithTaskEventsRecorderOnAssignedUserTask(TRACKED_EVENTS,
+                                                                                  null,
+                                                                                  TaskListener.EVENTNAME_ASSIGNMENT,
+                                                                                  CompletingTaskListener.class);
+    testRule.deploy(model);
+    runtimeService.startProcessInstanceByKey("process");
+    Task task = taskService.createTaskQuery().singleResult();
+
+    // when
+    taskService.setAssignee(task.getId(), "kermit");
+
+    // then
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    assertThat(orderedEvents.size()).isEqualTo(4);
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_UPDATE,
+                                              TaskListener.EVENTNAME_ASSIGNMENT,
+                                              TaskListener.EVENTNAME_COMPLETE);
+  }
+
+  @Test
+  public void testNoUpdateEventAfterCreateTaskListenerUpdatesProperties() {
```

Comment:
for new test classes we should try to stay consistent with the test method naming and stick to "shouldXYZ"

```suggestion
  public void shouldNotFireUpdateEventAfterCreateTaskListenerUpdatesProperties
```